source/common/http/conn_manager_impl.cc

```diff
@@ -1926,19 +1921,23 @@ void ConnectionManagerImpl::ActiveStream::setRoute(Router::RouteConstSharedPtr r
     return;
   }
 
-  filter_manager_.streamInfo().route_ = route;
-  setCachedRoute({std::move(route)});
-  if (nullptr == filter_manager_.streamInfo().route() ||
-      nullptr == filter_manager_.streamInfo().route()->routeEntry()) {
+  // Try to get the new cluster info based on the new route and set it to the
+  // cached_cluster_info_.
+  if (nullptr == route || nullptr == route->routeEntry()) {
     cached_cluster_info_ = nullptr;
   } else {
-    Upstream::ThreadLocalCluster* local_cluster =
-        connection_manager_.cluster_manager_.getThreadLocalCluster(
-            filter_manager_.streamInfo().route()->routeEntry()->clusterName());
-    cached_cluster_info_ = (nullptr == local_cluster) ? nullptr : local_cluster->info();
+    auto* cluster = connection_manager_.cluster_manager_.getThreadLocalCluster(
+        route->routeEntry()->clusterName());
+    cached_cluster_info_ = (nullptr == cluster) ? nullptr : cluster->info();
   }
 
+  // Update the cached route.
+  setCachedRoute({route});
+
+  // Update route and cluster info in the filter manager's stream info.
+  filter_manager_.streamInfo().route_ = std::move(route); // We can move route here.
   filter_manager_.streamInfo().setUpstreamClusterInfo(cached_cluster_info_.value());
+
```

Comment:
This is a more meaningful change. In previous code, We will set route the cached route and stream info. 

Then we will call the `filter_manager_.streamInfo().route()` to access the new route repeatly. Every call will create a temporary copy of shared pointer. (not sure if the compiler will optimize it or not)

The new code re-organize the logic to make it easy to read/understand (this point is more important) and eliminate all unnecessary copies.